dev/RadioButtons/RadioButtons.cpp

```diff
@@ -163,63 +661,214 @@ void RadioButtons::OnPropertyChanged(const winrt::DependencyPropertyChangedEvent
 
 void RadioButtons::UpdateItemsSource()
 {
-    if (auto listView = m_listView.get())
+    if (auto const repeater = m_repeater.get())
     {
-        if (ItemsSource())
+        m_itemsSourceChanged.revoke();
+        auto const source = GetItemsSource();
+
+        if (auto const item = repeater.TryGetElement(SelectedIndex()))
         {
-            listView.ItemsSource(ItemsSource());
+            if (auto const itemAsToggleButton = item.try_as<winrt::ToggleButton>())
+            {
+                itemAsToggleButton.IsChecked(false);
+            }
         }
-        else
+        m_selectionModel.ClearSelection();
+        repeater.ItemsSource(source);
+        m_selectionModel.Source(source);
+
+        //UpdateSelectionDPs([selectedIndexPath = m_selectionModel.SelectedIndex()]()
+        //    {
+        //        if (selectedIndexPath && selectedIndexPath.GetSize() == 1)
+        //        {
+        //            return selectedIndexPath.GetAt(0);
+        //        }
+        //        return -1;
+        //    }()
+        //);
+
+        if (auto const itemsSourceView = repeater.ItemsSourceView())
         {
-            listView.ItemsSource(Items());
+            m_itemsSourceChanged = itemsSourceView.CollectionChanged(winrt::auto_revoke, { this, &RadioButtons::OnRepeaterCollectionChanged });
         }
     }
 }
 
+winrt::IInspectable RadioButtons::GetItemsSource()
+{
+    if (auto const itemsSource = ItemsSource())
+    {
+        return itemsSource;
+    }
+    else
+    {
+        return Items();
+    }
+}
+
 void RadioButtons::UpdateMaximumColumns()
 {
-    if (auto itemsWrapGrid = m_itemsWrapGrid.get())
+    auto const maxColumns = MaximumColumns();
+    if (maxColumns <= 0)
     {
-        itemsWrapGrid.MaximumRowsOrColumns(MaximumColumns());
+        throw winrt::hresult_invalid_argument();
+    }
+
+    if (auto const repeater = m_repeater.get())
+    {
+        if (auto const layout = repeater.Layout().try_as<winrt::ColumnMajorUniformToLargestGridLayout>())
+        {
+            if (layout.MaximumColumns() != maxColumns)
+            {
+                layout.MaximumColumns(maxColumns);
+                repeater.InvalidateMeasure();
```

Comment:
the MaximumColumns property when changed in the layout, should call InvalidateMeasure itself. You should not have to call repeater's Invalidate measure here.

Reply:
Definitely

---
In reply to: [347509469](https://github.com/microsoft/microsoft-ui-xaml/pull/1623#discussion_r347509469) [](ancestors = 347509469)